mbus-api/src/mbus_nats.rs

```diff
@@ -0,0 +1,339 @@
+    }
+
+    /// The payload for mbus publish from the message.
+    fn payload(msg: &EventMessage) -> BusResult<bytes::Bytes> {
+        Ok(Bytes::from(serde_json::to_vec(msg)?))
+    }
+}
+
+#[async_trait]
+impl Bus for NatsMessageBus {
+    /// Publish messages to the message bus atmost once. The message is discarded if there is any
+    /// connection error after some retries.
+    async fn publish(&mut self, message: &EventMessage) -> BusResult<u64> {
+        let options = BackoffOptions::publish_backoff_options();
+        let mut tries = 0;
+        let mut log_error = true;
+
+        let subject = NatsMessageBus::subject(message);
+        let payload = NatsMessageBus::payload(message)?;
+
+        use std::time::Instant;
+        let now = Instant::now();
+
+        loop {
+            let publish_request = self.jetstream.publish(subject.clone(), payload.clone());
+            let err = match publish_request.await {
+                Ok(x) => match x.await {
+                    Ok(y) => {
+                        return Ok(y.sequence);
+                    }
+                    Err(error) => error,
+                },
+                Err(error) => error,
+            };
+
+            if log_error {
+                tracing::warn!(
+                    "Error publishing message to jetstream: {:?}. Retrying...",
+                    err
+                );
+                let _stream = self.verify_stream_exists().await;
+                log_error = false;
+            }
+
+            if tries == options.max_retries {
+                let elapsed = now.elapsed();
+                tracing::info!("Elapsed: {:.2?}", elapsed);
+
+                return Err(Error::PublishError {
+                    retries: options.max_retries,
+                    payload: format!("{message:?}"),
+                    error: err.to_string(),
+                });
+            }
+            if let Some(error) = err.downcast_ref::<std::io::Error>() {
+                if error.kind() == ErrorKind::TimedOut {
+                    tries += 1;
+                    continue;
+                }
+            }
+            backoff_with_options(&mut tries, &options).await;
+        }
+    }
+
+    /// Subscribes to the stream on messages though a consumer. Returns an ordered stream of
+    /// messages published to the stream.
+    async fn subscribe<T: Serialize + DeserializeOwned>(
+        &mut self,
+    ) -> BusResult<BusSubscription<T>> {
+        tracing::info!("Subscribing to Nats message bus");
+        let stream = self.get_or_create_stream(None).await?;
+
+        let messages = self.create_consumer_and_get_messages(stream).await?;
+        tracing::info!("Subscribed to Nats message bus successfully");
+
+        return Ok(BusSubscription {
+            messages,
+            _phantom: PhantomData::default(),
+        });
+    }
+}
+
+pub struct BusSubscription<T> {
```

Comment:
What does `T` represent here?

Reply:
The messages from bus are deserialised into the type T. T here is EventMessage since all the messages have same format.